pre_commit_hooks/markdown_links.py

```diff
@@ -0,0 +1,120 @@
+"""Library for loading links from a markdown file."""
+
+__copyright__ = """
+Copyright 2021 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import re
+from typing import Any, Dict, List, NamedTuple, Tuple
+
+import commonmark
+
+
+class Header(NamedTuple):
+    label: str
+    anchor: str
+    level: int
+
+
+class Link(NamedTuple):
+    label: str
+    destination: str
+    line_number: int
+
+
+def _make_label(node: Any) -> str:
+    """Makes a label from a node. Handles joining common children."""
+    label: List[str] = []
+    for child, entering in node.walker():
+        if child.t == "code":
+            label.extend(("`", child.literal, "`"))
+        elif child.t == "strong":
+            label.append("**")
+        elif child.t == "emph":
+            label.append("_")
+        elif child.t == "text":
+            label.append(child.literal)
+    return "".join(label)
+
+
+def _make_anchor(label: str, used_anchors: Dict[str, int]) -> str:
+    """Chooses the appropriate anchor name for a header label."""
+    anchor = label.lower().strip()
+    # Imitate GFM anchors. Sadly, the exact format isn't clearly documented
+    # anywhere, so this based on experimentation and what others have done to
+    # successfully replicate the GitHub anchor mapping.
+    anchor = anchor.replace(" ", "-")
+    anchor = re.sub("[!\"#$%&'()*+,./:;<=>?@[\\\\\\]^`{|}~]", "", anchor)
```

Comment:
I see this is just code moving around, so not blocking and fine for a follow-up: hoist to a precomputed refer global since this is in the loop over the file?

Reply:
Well, no noticeable performance hit -- but this is Python, this probably isn't significant cost.